File: frontend/js/views/markup_margin.js
// Markup vs margin — from cost + one of {price, markup%, margin%}, shows
// price, profit, and both markup% (of cost) and margin% (of price), via
// /calc/markup-margin. Updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const pct = (n) => Number(n).toFixed(2) + '%';
const VIEW = 'markup-margin';
let lastReport = null;
let lastBody = null;

export async function renderMarkupMargin(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.mm.h1.title">// MARKUP vs MARGIN</span></h1>
        <p class="muted small" data-i18n="view.mm.hint.intro">
            Markup is profit as a percent of cost; margin is the same profit as a percent of
            price. They are never equal — a 50% markup is only a 33% margin, and doubling cost
            (100% markup, "keystone") is a 50% margin. Pricing off the wrong one erodes profit.
            Pick what you know — price, markup%, or margin% — and see all four. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.mm.h2.inputs">What you know</h2>
            <form id="mm-form" class="inline-form">
                <label><span data-i18n="view.mm.label.mode">I'm entering</span>
                    <select name="mode" id="mm-mode">
                        <option value="price" data-i18n="view.mm.mode.price">Cost + price</option>
                        <option value="markup" data-i18n="view.mm.mode.markup">Cost + markup %</option>
                        <option value="margin" data-i18n="view.mm.mode.margin">Cost + margin %</option>
                    </select></label>
                <label><span data-i18n="view.mm.label.cost">Unit cost ($)</span>
                    <input type="number" step="0.01" min="0" name="cost_usd" value="60" required></label>
                <label id="mm-price-wrap"><span data-i18n="view.mm.label.price">Selling price ($)</span>
                    <input type="number" step="0.01" min="0" name="price_usd" value="100"></label>
                <label id="mm-markup-wrap"><span data-i18n="view.mm.label.markup">Markup (%)</span>
                    <input type="number" step="0.01" name="markup_pct" value="66.67"></label>
                <label id="mm-margin-wrap"><span data-i18n="view.mm.label.margin">Margin (%)</span>
                    <input type="number" step="0.01" name="margin_pct" value="40"></label>
            </form>
            <div id="mm-tools" class="ce-toolbar"></div>
            <button type="button" id="mm-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="mm-sens" class="ce-sens"></div>
        </div>
        <div id="mm-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#mm-form');
    const modeSel = mount.querySelector('#mm-mode');
    const wraps = {
        price: mount.querySelector('#mm-price-wrap'),
        markup: mount.querySelector('#mm-markup-wrap'),
        margin: mount.querySelector('#mm-margin-wrap'),
    };
    const syncMode = () => {
        for (const [k, el] of Object.entries(wraps)) el.style.display = modeSel.value === k ? '' : 'none';
    };
    enh.prefillForm(form, enh.readHashInputs());

    const readBody = () => {
        const fd = new FormData(form);
        return {
            mode: fd.get('mode'),
            cost_usd: Number(fd.get('cost_usd')) || 0,
            price_usd: Number(fd.get('price_usd')) || 0,
            markup_pct: Number(fd.get('markup_pct')) || 0,
            margin_pct: Number(fd.get('margin_pct')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcMarkupMargin(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.mm.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    enh.mountToolbar(mount.querySelector('#mm-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'markup-margin.csv' });
    mount.querySelector('#mm-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    modeSel.addEventListener('change', () => { syncMode(); generate(); });
    form.addEventListener('input', live);
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    syncMode();
    generate();
}

function reportRows(r) {
    if (!r || !r.feasible) return [];
    return [
        ['metric', 'value'],
        ['price_usd', r.price_usd],
        ['profit_usd', r.profit_usd],
        ['markup_pct', r.markup_pct],
        ['margin_pct', r.margin_pct],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#mm-result');
    if (!r.feasible) {
        el.innerHTML = `<div class="chart-panel"><h2 data-i18n="view.mm.h2.result">The numbers</h2>
            <p class="muted small neg" data-i18n="view.mm.warn.infeasible">A margin of 100% or more is impossible — price would be infinite. Lower the margin.</p></div>`;
        applyUiI18n(el);
        return;
    }
    const profitCls = r.profit_usd >= 0 ? 'pos' : 'neg';
    // Line chart: the canonical markup→margin curve (margin % vs markup % from 0 → 200%),
    // independent of the chosen input mode — it is a pure relationship.
    const xs = enh.linspace(0, 200, 21);
    const pts = await Promise.all(xs.map(async (mk) => {
        const rr = await api.calcMarkupMargin({ mode: 'markup', cost_usd: body.cost_usd || 1, price_usd: 0, markup_pct: mk, margin_pct: 0 });
        return { x: mk, y: rr && rr.feasible ? rr.margin_pct : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'markup %', ylabel: 'margin %' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.mm.h2.result">The numbers</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.mm.card.price">Price</div>
                    <div class="value">${money(r.price_usd)}</div></div>
                <div class="card ${profitCls}"><div class="label" data-i18n="view.mm.card.profit">Profit / unit</div>
                    <div class="value ${profitCls}">${money(r.profit_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.mm.card.markup">Markup (of cost)</div>
                    <div class="value">${pct(r.markup_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.mm.card.margin">Margin (of price)</div>
                    <div class="value">${pct(r.margin_pct)}</div></div>
            </div>
            ${chart}
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#mm-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: unit cost 0.5× → 1.5×; y: markup 0 → 150%. Output: profit per unit (in markup mode).
    const cost = base.cost_usd || 60;
    const xVals = enh.linspace(cost * 0.5, cost * 1.5, 5);
    const yVals = enh.linspace(0, 150, 5);
    const { cells } = await enh.runSensitivity({ base: { mode: 'markup', cost_usd: cost, price_usd: 0, markup_pct: 0, margin_pct: 0 }, xKey: 'cost_usd', yKey: 'markup_pct', xVals, yVals, compute: (b) => api.calcMarkupMargin(b), pick: (r) => (r && r.feasible ? r.profit_usd : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells, fmt: (v) => (v == null ? '—' : '$' + v.toFixed(0)), xfmt: (v) => '$' + v.toFixed(0), yfmt: (v) => v.toFixed(0) + '%', xName: t('view.mm.label.cost') || 'Cost', yName: t('view.mm.label.markup') || 'Markup' });
}

File: crates/traderview-core/src/markup_margin.rs
//! Markup vs margin — the conversion small businesses get wrong.
//!
//! **Markup** is profit as a percent of **cost**; **margin** is the same
//! profit as a percent of **price**. They are never equal (except at zero):
//! a 50% markup on a $50 cost is a $75 price — but that's only a 33% margin.
//! Pricing off the wrong one quietly erodes profit.
//!
//! Given the unit cost and one of {price, markup%, margin%}, this returns
//! all four figures so the two can be compared directly:
//!
//!   * markup% = profit / cost × 100
//!   * margin% = profit / price × 100
//!   * from markup: price = cost × (1 + markup/100)
//!   * from margin: price = cost / (1 − margin/100)   (margin must be < 100)
//!
//! Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Mode {
    /// Cost + price are given.
    Price,
    /// Cost + markup% are given.
    Markup,
    /// Cost + margin% are given.
    Margin,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MarkupInput {
    pub mode: Mode,
    pub cost_usd: f64,
    /// Selling price (Price mode).
    #[serde(default)]
    pub price_usd: f64,
    /// Markup percent of cost (Markup mode).
    #[serde(default)]
    pub markup_pct: f64,
    /// Margin percent of price (Margin mode).
    #[serde(default)]
    pub margin_pct: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct MarkupResult {
    pub cost_usd: f64,
    pub price_usd: f64,
    pub profit_usd: f64,
    pub markup_pct: f64,
    pub margin_pct: f64,
    /// False when the inputs are infeasible (e.g. margin ≥ 100%).
    pub feasible: bool,
}

pub fn analyze(i: &MarkupInput) -> MarkupResult {
    let cost = i.cost_usd.max(0.0);
    let mut feasible = true;

    let price = match i.mode {
        Mode::Price => i.price_usd.max(0.0),
        Mode::Markup => cost * (1.0 + i.markup_pct / 100.0),
        Mode::Margin => {
            // price = cost / (1 − margin); margin ≥ 100% is impossible.
            if i.margin_pct >= 100.0 {
                feasible = false;
                0.0
            } else {
                cost / (1.0 - i.margin_pct / 100.0)
            }
        }
    };

    let profit = price - cost;
    let markup_pct = if cost > 0.0 { profit / cost * 100.0 } else { 0.0 };
    let margin_pct = if price > 0.0 { profit / price * 100.0 } else { 0.0 };

    MarkupResult {
        cost_usd: cost,
        price_usd: price,
        profit_usd: profit,
        markup_pct,
        margin_pct,
        feasible,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn price_mode_computes_markup_and_margin() {
        // cost 60, price 100 → profit 40, markup 66.67%, margin 40%.
        let r = analyze(&MarkupInput {
            mode: Mode::Price,
            cost_usd: 60.0,
            price_usd: 100.0,
            markup_pct: 0.0,
            margin_pct: 0.0,
        });
        assert!((r.profit_usd - 40.0).abs() < 1e-9);
        assert!((r.markup_pct - (40.0 / 60.0 * 100.0)).abs() < 1e-9);
        assert!((r.margin_pct - 40.0).abs() < 1e-9);
    }

    #[test]
    fn markup_mode_recovers_price() {
        // cost 60, markup 66.666…% → price 100.
        let r = analyze(&MarkupInput {
            mode: Mode::Markup,
            cost_usd: 60.0,
            price_usd: 0.0,
            markup_pct: 40.0 / 60.0 * 100.0,
            margin_pct: 0.0,
        });
        assert!((r.price_usd - 100.0).abs() < 1e-9);
        assert!((r.margin_pct - 40.0).abs() < 1e-9);
    }

    #[test]
    fn margin_mode_recovers_price() {
        // cost 60, margin 40% → price = 60 / 0.6 = 100.
        let r = analyze(&MarkupInput {
            mode: Mode::Margin,
            cost_usd: 60.0,
            price_usd: 0.0,
            markup_pct: 0.0,
            margin_pct: 40.0,
        });
        assert!((r.price_usd - 100.0).abs() < 1e-9);
        assert!((r.markup_pct - (40.0 / 60.0 * 100.0)).abs() < 1e-9);
        assert!(r.feasible);
    }

    #[test]
    fn markup_never_equals_margin() {
        // cost 50, price 100 → markup 100%, margin 50%.
        let r = analyze(&MarkupInput {
            mode: Mode::Price,
            cost_usd: 50.0,
            price_usd: 100.0,
            markup_pct: 0.0,
            margin_pct: 0.0,
        });
        assert!((r.markup_pct - 100.0).abs() < 1e-9);
        assert!((r.margin_pct - 50.0).abs() < 1e-9);
    }

    #[test]
    fn margin_at_or_above_100_is_infeasible() {
        let r = analyze(&MarkupInput {
            mode: Mode::Margin,
            cost_usd: 60.0,
            price_usd: 0.0,
            markup_pct: 0.0,
            margin_pct: 100.0,
        });
        assert!(!r.feasible);
        assert!(r.price_usd.abs() < 1e-9);
    }

    #[test]
    fn zero_cost_guards_markup() {
        let r = analyze(&MarkupInput {
            mode: Mode::Price,
            cost_usd: 0.0,
            price_usd: 100.0,
            markup_pct: 0.0,
            margin_pct: 0.0,
        });
        assert!(r.markup_pct.abs() < 1e-9);
        assert!((r.margin_pct - 100.0).abs() < 1e-9); // all profit
    }

    #[test]
    fn markup_to_margin_roundtrip() {
        // From a 25% markup, derive price+margin, then re-derive markup.
        let a = analyze(&MarkupInput {
            mode: Mode::Markup,
            cost_usd: 80.0,
            price_usd: 0.0,
            markup_pct: 25.0,
            margin_pct: 0.0,
        });
        let b = analyze(&MarkupInput {
            mode: Mode::Margin,
            cost_usd: 80.0,
            price_usd: 0.0,
            markup_pct: 0.0,
            margin_pct: a.margin_pct,
        });
        assert!((b.markup_pct - 25.0).abs() < 1e-6);
        assert!((b.price_usd - a.price_usd).abs() < 1e-6);
    }

    #[test]
    fn keystone_double_is_fifty_percent_margin() {
        // Keystone pricing = 100% markup (double cost) → exactly 50% margin.
        let r = analyze(&MarkupInput {
            mode: Mode::Markup,
            cost_usd: 25.0,
            price_usd: 0.0,
            markup_pct: 100.0,
            margin_pct: 0.0,
        });
        assert!((r.price_usd - 50.0).abs() < 1e-9);
        assert!((r.margin_pct - 50.0).abs() < 1e-9);
    }
}
